### ISSUE TITLE
Determine how sources should be created / associated with accounts (and account actions)

### ISSUE THREAD

**Issue by slifty** [PhilanthropyDataCommons/service#1146]
We're about to support provenance in our data structure / endpoints which is great (#1116 / #1094), but now we need to decide how sources get created and how a given API call or bulk upload gets associated with a source.

Some possible approaches:

1. accounts are linked to a specific source, and that account's actions are then associated with that source.
2. Accounts are linked to multiple sources, and that account's actions CAN then be associated with various sources (and the selected source would need to be provided as part of the API call).
3. Accounts are linked to one (or more) organizations / funders / data_providers, and the source(s) associated with those entities are then assigned to the account.

Regardless we will still need to think about how and when an account association with a source (or organizational entity) is done.  Also, whether that association should be 1:n or n:m.

**Comment by slifty** [comment]
I spoke with @jasonaowen about this and we think we have an approach.

1. add a `sourceId` column to `accounts` (nullable).
2. Use the sourceId when creating proposal versions.
3. Add a `PUT /accounts` endpoint
4. If an account doesn't have a source ID, it won't be able to create proposal versions or bulk uploads.

We'll also adjust accounts to use GUID as a natural key -- authentication Id is assigned by keycloak and can't change.  This should make it easier for admins to create accounts and associate sources.  This also means we don't need to create a `POST` endpoint.

---

What this means is that account creation will require an administrator to assign a source to any new account that would have write permissions.

**Comment by bickelj** [comment]
I am late to the party now that PR #1191 is already created. I am surprised at the association of a source with an account but I see above that there is good reason to associate them. To have better feedback, I need to look at the current state of provenance. My initial reaction is that "who" is separate from "where" and that provenance is supposed to capture both. And I think that expectation is already included in the discussion above as well, thus my need to take a closer look.

**Comment by slifty** [comment]
@bickelj the short story is that provenance is the "where" at this point, but ultimately the system needs to be able to associate the "where" to an action somehow.

From a UX perspective at this point in the project the shortest path to implementation is to just have a "where" linked to an account -- so when an account takes an action, we know the "where" and can make that association automatically without any need for user input.

Eventually we will want to consider having a more complex process by which a user specifies the where in their API call, AND the system is able to validate whether that user is allowed to specify that where (we don't want someone saying that their data came from Candid if they are not vetted as being an agent of Candid).

TL;DR I expect `sourceId` to be removed from `user` eventually and replaced with a many-to-many `userSourceAgentRelationships` table or some such thing.

**Comment by slifty** [comment]
@bickelj That would not provide access to a source though / would bypass the provenance system.

**Comment by bickelj** [comment]
@slifty Yeah, my assumption is that the caller, at least for now, declares the source explicitly. If that caller has write access, that caller has some presumed level of trust. The more complex thing you mentioned, namely limiting who can declare what source, is more complex and can come later.

**Comment by bickelj** [comment]
@slifty My fear is automatically-created inaccurate data. If a caller writes the wrong source, I see that as different. If we automatically generate a source, we want to be dang sure it's correct.

**Comment by slifty** [comment]
> Yeah, my assumption is that the caller, at least for now, declares the source explicitly. If that caller has write access, that caller has some presumed level of trust. The more complex thing you mentioned, namely limiting who can declare what source, is more complex and can come later.

@bickelj You're right -- that's a good argument and has less code byproducts / tech debt associated with its implementation.

I want to confirm with @kfogel that we're OK with that level of user trust for now, but if we are then I would ditch the open PR that makes the user association and instead modify the API call to require the user to provide a sourceId when adding data.

------

> My fear is automatically-created inaccurate data. If a caller writes the wrong source, I see that as different. If we automatically generate a source, we want to be dang sure it's correct.

I was going to say "the user source association would be manually set" but then realized that you're pointing out that a user might upload data that came from another source without realizing that their account would passively link that data to something else (aka someone may have a Ford Foundation account, but upload some data that came from somewhere else, and now because of system magic that data is falsely attributed to Ford Foundation)

**Comment by slifty** [comment]
Thank you for raising all this @bickelj!  I'm pretty solidly convinced that we should ditch this path and make a required parameter -- and we can augment the system with checks / authorization at a later date when that added layer of protection has been authorized.

**Comment by bickelj** [comment]
@slifty said:
> I was going to say "the user source association would be manually set" but then realized that you're pointing out that a user might upload data that came from another source without realizing that their account would passively link that data to something else (aka someone may have a Ford Foundation account, but upload some data that came from somewhere else, and now because of system magic that data is falsely attributed to Ford Foundation)

Yes, exactly.

My perspective, of course, comes from being a technical member of the PDC team having posted data from many sources.
